urld0f681c558_farom57_AstroidDriver_tgz-pJ8-laazotea_indi_server_examples_MinimumINDIServerJ8/src/laazotea/indi/Base64.java

```diff
@@ -0,0 +1,1977 @@
+      return new byte[0];
+    } else if (len < 4) {
+      throw new IllegalArgumentException(
+              "Base64-encoded string must have at least four characters, but length specified was " + len);
+    }   // end if
+
+    byte[] DECODABET = getDecodabet(options);
+
+    int len34 = len * 3 / 4;       // Estimate on array size
+    byte[] outBuff = new byte[len34]; // Upper limit on size of output
+    int outBuffPosn = 0;             // Keep track of where we're writing
+
+    byte[] b4 = new byte[4];     // Four byte buffer from source, eliminating white space
+    int b4Posn = 0;               // Keep track of four byte input buffer
+    int i = 0;               // Source array counter
+    byte sbiDecode = 0;               // Special value from DECODABET
+
+    for (i = off ; i < off + len ; i++) {  // Loop through source
+
+      sbiDecode = DECODABET[ source[i] & 0xFF];
+
+      // White space, Equals sign, or legit Base64 character
+      // Note the values such as -5 and -9 in the
+      // DECODABETs at the top of the file.
+      if (sbiDecode >= WHITE_SPACE_ENC) {
+        if (sbiDecode >= EQUALS_SIGN_ENC) {
+          b4[ b4Posn++] = source[i];         // Save non-whitespace
+          if (b4Posn > 3) {                  // Time to decode?
+            outBuffPosn += decode4to3(b4, 0, outBuff, outBuffPosn, options);
+            b4Posn = 0;
+
+            // If that was the equals sign, break out of 'for' loop
+            if (source[i] == EQUALS_SIGN) {
+              break;
+            }   // end if: equals sign
+          }   // end if: quartet built
+        }   // end if: equals sign or better
+      } // end if: white space, equals sign or better
+      else {
+        // There's a bad input character in the Base64 stream.
+        throw new java.io.IOException(String.format(
+                "Bad Base64 input character decimal %d in array position %d", ((int) source[i]) & 0xFF, i));
+      }   // end else: 
+    }   // each input character
+
+    byte[] out = new byte[outBuffPosn];
+    System.arraycopy(outBuff, 0, out, 0, outBuffPosn);
+    return out;
+  }   // end decode
+
+  /**
+   * Decodes data from Base64 notation, automatically detecting gzip-compressed
+   * data and decompressing it.
+   *
+   * @param s the string to decode
+   * @return the decoded data
+   * @throws java.io.IOException If there is a problem
+   * @since 1.4
+   */
+  public static byte[] decode(String s) throws java.io.IOException {
+    return decode(s, NO_OPTIONS);
+  }
+
+  /**
+   * Decodes data from Base64 notation, automatically detecting gzip-compressed
+   * data and decompressing it.
+   *
+   * @param s the string to decode
+   * @param options encode options such as URL_SAFE
+   * @return the decoded data
+   * @throws java.io.IOException if there is an error
+   * @throws NullPointerException if <tt>s</tt> is null
+   * @since 1.4
+   */
+  public static byte[] decode(String s, int options) throws java.io.IOException {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 14. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.